akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -239,43 +239,49 @@ private[akka] final case class Fold[In, Out](zero: Out, f: (Out, In) ⇒ Out, de
 /**
  * INTERNAL API
  */
-private[akka] final case class Intersperse[T](start: Option[T], inject: T, end: Option[T]) extends StatefulStage[T, T] {
-  private var needsToEmitStart = start.isDefined
+final case class Intersperse[T](start: Option[T], inject: T, end: Option[T]) extends GraphStage[FlowShape[T, T]] {
 
-  override def initial: StageState[T, T] =
-    start match {
-      case Some(initial) ⇒ firstWithInitial(initial)
-      case _             ⇒ first
-    }
+  private val in = Inlet[T]("in")
+  private val out = Outlet[T]("out")
 
-  def firstWithInitial(initial: T) = new StageState[T, T] {
-    override def onPush(elem: T, ctx: Context[T]) = {
-      needsToEmitStart = false
-      emit(Iterator(initial, elem), ctx, running)
-    }
-  }
+  override val shape = FlowShape(in, out)
 
-  def first = new StageState[T, T] {
-    override def onPush(elem: T, ctx: Context[T]) = {
-      become(running)
-      ctx.push(elem)
-    }
-  }
+  override def createLogic(attr: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    var first = true // is this the first element?
 
-  def running = new StageState[T, T] {
-    override def onPush(elem: T, ctx: Context[T]): SyncDirective =
-      emit(Iterator(inject, elem), ctx)
-  }
+    setHandler(in, new InHandler {
+      override def onPush(): Unit = {
+        if (first) {
+          first = false
+          // emit: start if defined, elem
+          if (start.isDefined) emitMultipe(out, Iterator(start.get, grab(in)))
```

Comment:
@lolski we've fixed a typo here, the method is now `emitMultiple`, could you rebase this PR please?